### PR TITLE
i#5675 record filter: Improve interfaces.

### PR DESCRIPTION
**Pull request by abhinav92003**
Improves interfaces to allow use in 3rd party code.

Uses std::unique_ptr for the record_filter_func_t vector and for the input file in record_file_reader_t. Adjusts record_filter_launcher and record_filter_unit_tests to pass unique_ptrs now.

Make input_entry_count_ and output_entry_count_ protected to let them be accessed by derived classes.

Separates out the cache_filter implementation details into a cpp file.

Adds the drcachesim/simulator directory to includes.

Issue: #5675

### CODE REVIEW

**Comment by abhinav92003** [comment]
x86-32 failure is  #4604.

File: clients/drcachesim/tools/record_filter_launcher.cpp [hunk 1]
```diff
@@ -40,6 +40,8 @@
 #include "tools/filter/type_filter.h"
 #include "tools/filter/record_filter.h"
 
+#include <limits>
```
Comment (derekbruening) [inline clients/drcachesim/tools/record_filter_launcher.cpp:43]: This looks unrelated to the other changes in this file

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/reader/compressed_file_reader.cpp [hunk 2]
```diff
@@ -147,15 +146,15 @@ record_file_reader_t<gzip_reader_t>::open_single_file(const std::string &path)
     if (!open_single_file_common(path, file))
         return false;
     VPRINT(this, 1, "Opened input file %s\n", path.c_str());
-    input_file_ = new gzip_reader_t(file);
+    input_file_ = std::unique_ptr<gzip_reader_t>(new gzip_reader_t(file));
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/reader/compressed_file_reader.cpp:149]: It needs to be unless we provide a default constructor for all possible template parameters.

File: clients/drcachesim/tools/record_filter_launcher.cpp (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/record_filter_launcher.cpp:43]: It complains of missing stuff in `std::numeric_limits<uint64_t>::max()` below. Maybe some transitive dependency via cache_filter.h included them before.

**Comment by abhinav92003** [comment]
Jenkins failure is #5635.

**Comment by abhinav92003** [comment]
run arm tests

**Comment by abhinav92003** [comment]
Jenkins now hit #5484.

**Comment by abhinav92003** [comment]
run arm tests
